src/node_v8.cc

```diff
@@ -157,6 +157,21 @@ void CachedDataVersionTag(const FunctionCallbackInfo<Value>& args) {
   args.GetReturnValue().Set(result);
 }
 
+void SetHeapSnapshotNearHeapLimit(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args[0]->IsNumber());
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* const isolate = args.GetIsolate();
+  int64_t limit = args[0].As<v8::Number>()->Value();
+  CHECK_GT(limit, 0);
+  // Do not set twice which will make the process crash.
+  if (!env->near_heap_callback_is_added()) {
+    env->set_near_heap_callback_is_added(true);
+    isolate->AddNearHeapLimitCallback(Environment::NearHeapLimitCallback, env);
+  }
+  env->set_heap_snapshot_near_heap_limit(limit);
+  env->set_heap_limit_snapshot_taken(0);
```

Comment:
I don't think this reset should be performed in this procedure. If people need to create more heap snapshots near the heap limit, they can simply increase the limit if applicable.

Reply:
I'd find it surprising that updating the limit resets the counter too...It also seems simply easier to maintain the invariant that `env->heap_snapshot_near_heap_limit() > 0` means the near heap limit callback for heap snapshots is configured.

Reply:
If at the beginning, the user sets the limit to 3, and after 3 snapshots are taken( limit: 3, taken:3 ), the user set s the limit to 2, we only can taken one more snapshot, because 2 < 3 so we will remove the near limit callback once the near limit callback is called again. Maybe we just reset the value of taken when limit < taken ?

Reply:
And i have another question. The code as follows is called when the `Environment` is destructed.
```
if (options_->heap_snapshot_near_heap_limit > heap_limit_snapshot_taken_) {
    isolate_->RemoveNearHeapLimitCallback(Environment::NearHeapLimitCallback, 0);
}
```
But we will call `RemoveNearHeapLimitCallback` early in near limit callback if `estimated_overhead > available` is true. The code is as follows.
```
 if (estimated_overhead > available) {
    env->isolate()->RemoveNearHeapLimitCallback(NearHeapLimitCallback, initial_heap_limit);
    return current_heap_limit + 1;
  }
```
This looks like a problem which `RemoveNearHeapLimitCallback` is called twice. So i add the `near_heap_callback_is_added` to track whether the near limit callback is added.

Reply:
> the user set s the limit to 2, we only can taken one more snapshot, because 2 < 3 so we will remove the near limit callback once the near limit callback is called again. 

I think we can just throw an error or an warning in that case

> This looks like a problem which RemoveNearHeapLimitCallback is called twice.

It's a no-op from V8 if the callback is already removed, so I think that'll be fine?

Reply:
Or we can simply error from the API if the limit is already configured from the CLI - in general we don't allow overriding configurations from CLI in the JS APIs, after all.

Reply:
From the test on v18.8.0, adding or removing the same callback twice will crash the process. The code of V8 is as follows.
```
void Heap::AddNearHeapLimitCallback(v8::NearHeapLimitCallback callback,
                                    void* data) {
  const size_t kMaxCallbacks = 100;
  CHECK_LT(near_heap_limit_callbacks_.size(), kMaxCallbacks);
  for (auto callback_data : near_heap_limit_callbacks_) {
    CHECK_NE(callback_data.first, callback);
  }
  near_heap_limit_callbacks_.push_back(std::make_pair(callback, data));
}


void Heap::RemoveNearHeapLimitCallback(v8::NearHeapLimitCallback callback,
                                       size_t heap_limit) {
  for (size_t i = 0; i < near_heap_limit_callbacks_.size(); i++) {
    if (near_heap_limit_callbacks_[i].first == callback) {
      near_heap_limit_callbacks_.erase(near_heap_limit_callbacks_.begin() + i);
      if (heap_limit) {
        RestoreHeapLimit(heap_limit);
      }
      return;
    }
  }
  UNREACHABLE();
}
```

Reply:
Oh, right, I think in that case we can simply remove the `RemoveNearHeapLimitCallback()` call from `Environment::NearHeapLimitCallback` - I don't recall why it was added, but it seems unnecessary to me.

Reply:
Wait..I think we do need to keep the `RemoveNearHeapLimitCallback()` call, so that the worker callback can remain to throw the error if we never manage to taken enough heap snapshots. Then I think yes we do need a state to track the callback for heapsnapshots, though the name of the field needs to indicate that it is only about `Environment::NearHeapLimitCallback`, not about the presence of near heap limit callbacks overall (because there's the worker one which will always be there for worker environments), perhaps `env->heapsnapshot_near_heap_limit_callback_added()`?

Reply:
Thanks.
1. I will rename `near_heap_limit_callback_is_added` to `heapsnapshot_near_heap_limit_callback_added`.   
2. I think we can throw an error or a warning if the limit set by JS API is less than the current value of limit and greater than the current value of taken.
3. Will we allow the user to reset the limit by JS API if the limit is already configured from the CLI ?

What do you think? @joyeecheung @legendecas

Reply:
1 STGM. Regarding 2 or 3, I think the saner choice would be to only allow the limit to be configured once - either from the CLI, or from the JS API, it can be error-prone to allow too much mutation from within JS which can itself trigger allocation. If the application actually reaches the heap limit and writes heapsnapshots to disk, it's probably already in a pretty bad shape, if it just keeps trying raising the limit that can be a can of worms..

Reply:
I have updated the code, please help review again, thanks !